### PR TITLE
Add an interface to create Context with CancellationToken

### PR DESCRIPTION
**Pull request by BoyaWu10**
Add an interface to create Context with CancellationToken

### CODE REVIEW

**Comment by BoyaWu10** [comment]
@daviburg Could you please help to review this PR?

File: src/DotLiquid.Tests/ExceptionHandlingTests.cs [hunk 1]
```diff
@@ -103,6 +105,19 @@ public void TestTimeoutError()
             });
         }
 
+        [Test]
+        public void TestOperationCancelledError()
+        {
+            var template = Template.Parse(" {% for i in (1..1000000) %} {{ i }} {% endfor %} ");
+            var source = new CancellationTokenSource(100);
+            var context = new Context(new List<Hash>(), new Hash(), new Hash(), ErrorsOutputMode.Rethrow, 0, source.Token, CultureInfo.InvariantCulture);
```
Comment (daviburg) [inline src/DotLiquid.Tests/ExceptionHandlingTests.cs:113]: I recommend to use named parameters when using constant input values such as 0 or null so it is easier to understand for human readers.

File: src/DotLiquid/Context.cs [hunk 2]
```diff
@@ -91,6 +93,39 @@ public Context
...
+        /// <param name="outerScope"></param>
+        /// <param name="registers"></param>
+        /// <param name="errorsOutputMode"></param>
+        public Context
+            (List<Hash> environments
+             , Hash outerScope
+             , Hash registers
+             , ErrorsOutputMode errorsOutputMode
+             , int maxIterations
+             , CancellationToken cancellationToken
```
Comment (daviburg) [inline src/DotLiquid/Context.cs:109]: Why not add at the end?

File: src/DotLiquid/Context.cs [hunk 3]
```diff
@@ -91,6 +93,39 @@ public Context
             SquashInstanceAssignsWithEnvironments();
         }
 
+        /// <summary>
+        /// Creates a new rendering context
+        /// </summary>
+        /// <param name="environments"></param>
+        /// <param name="outerScope"></param>
+        /// <param name="registers"></param>
+        /// <param name="errorsOutputMode"></param>
```
Comment (daviburg) [inline src/DotLiquid/Context.cs:102]: Add the XML comment for the new parameter and take the opportunity to fill in the missing comments content for the existing parameters.

File: src/DotLiquid/Context.cs [hunk 4]
```diff
@@ -91,6 +93,39 @@ public Context
             SquashInstanceAssignsWithEnvironments();
```
Comment (daviburg) [inline src/DotLiquid/Context.cs:93]: Avoid code duplication by having one of the two methods call the other one instead.

**Review by daviburg — CHANGES_REQUESTED** [review]

**Comment by daviburg** [comment]
Please add a feature work item to the backlog of this and link it to this PR

**Comment by BoyaWu10** [comment]
Thanks for the comments. They are fixed in the update. Do we track feature work items in GitHub issues? If so, I've created an issue #407 for this PR.

**Comment by daviburg** [comment]
> Thanks for the comments. They are fixed in the update. Do we track feature work items in GitHub issues? If so, I've created an issue #407 for this PR.

As we have multiple companies wanting to add various features to DotLiquid lately, we should be a bit more formal with the changes, so yes, we should have a item for each change. I'm not a GitHub expert, but it may well be that all items are 'issues' even feature requests.

File: src/DotLiquid/Context.cs (hunk 3 above)
Comment (daviburg) [reply src/DotLiquid/Context.cs:91]: The XML param declaration has been added but the comment is still empty. Each comment should be something like:

<param name="errorsOutputMode">Errors output mode</param>

(Feel free to expand the descriptions based on your knowledge of what the parameter does).

**Review by daviburg — APPROVED** [review]

**Pull request merged**